### PR TITLE
Add signal to hide regform sections from plugins

### PR DESCRIPTION
**Pull request by OmeGak**

### CODE REVIEW

File: indico/core/signals/event/registration.py [hunk 1]
```diff
@@ -98,3 +98,14 @@
 The list of badge templates is passed in the `badge_templates` kwarg.
 The signal handler is expected to mutate the list.
 """)
+
+registration_form_section_can_access = _signals.signal('registration-form-section-can-render', """
```
Comment (ThiefMaster) [inline indico/core/signals/event/registration.py:102]: `can_access` vs `can-render`

File: indico/core/signals/event/registration.py [hunk 2]
```diff
@@ -98,3 +98,14 @@
...
+Called when generating registration form sections data.  The `sender` is the
+``RegistrationForm`` object.  The section object is passed in the `section`
+kwarg the registration object iss be passed in the `registration` kwarg.  If the
```
Comment (ThiefMaster) [inline indico/core/signals/event/registration.py:105]: "iss be passed"? 🙈

File: indico/core/signals/event/registration.py [hunk 3]
```diff
@@ -98,3 +98,14 @@
...
+access check fails, the section will be skipped.
+
+If the signal returns ``True`` or ``False``, the access check succeeds
+or fails immediately.  If multiple subscribers to the signal return
```
Comment (ThiefMaster) [inline indico/core/signals/event/registration.py:109]: This first sentence is a bit misleading since the check does not short-circuit.

**Comment by ThiefMaster** [comment]
What happens if you register and then edit the registration with a section being hidden by that signal?

Would the data be lost or simply not touched? I think the latter would be the correct behavior.

In `modify_registration` we have this logic for manager-only sections - I think you may need something similar for anything hidden by the signal:

```python
        if management or not form_item.parent.is_manager_only:
            value = data.get(form_item.html_field_name)
```

If you add another check to that condition, you also automatically ignore data that may have been submitted even though the section isn't visible. But ideally you'd also not add any fields to the WTForm in `make_registration_form` if the section gets hidden.

File: indico/modules/events/registration/util.py [hunk 4]
```diff
@@ -280,6 +288,7 @@ def modify_registration(registration, data, management=False, notify_user=True):
             if getattr(registration, key) != value:
                 personal_data_changes[key] = value
             setattr(registration, key, value)
+    raise Exception
```
Comment (ThiefMaster) [inline indico/modules/events/registration/util.py:291]: 💣

File: indico/modules/events/registration/util.py [hunk 5]
```diff
@@ -158,7 +163,8 @@ def validate_email(self, field):
     for form_item in regform.active_fields:
         if not management and form_item.parent.is_manager_only:
             continue
-
+        if not all(values_from_signal(signal.send(regform, section=form_item.parent, registration=registration))):
```
Comment (ThiefMaster) [inline indico/modules/events/registration/util.py:166]: You use it in 3 places and it gets called for each form item. What about moving it to a util like this?

```python
@memoize_request
def _is_section_visible(regform, section, registration=None):
    return all(values_from_signal(signal.send(regform, section=section, registration=registration)))
```

That way you don't spam the signal that much in larger registration forms.

**Review by ThiefMaster — DISMISSED** [review]

**Comment by ThiefMaster** [comment]
I just tested it, any I found some problems:

- I still see a section which is being hidden by the plugin when registering.
- After registering, I added another field to the regform. When editing the existing registration, I got the KeyError below (doesn't happen if the plugin doesn't hide the section)

```
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/util.py", line 301, in modify_registration
    notify_registration_modification(registration, notify_user)
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/notifications.py", line 60, in notify_registration_modification
    _notify_registration(registration, 'registration_modification_to_registrant.html')
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/notifications.py", line 45, in _notify_registration
    mail = make_email(to_list=to_list, template=template, html=True, from_address=from_address, attachments=attachments)
  File "/home/adrian/dev/indico/src/indico/core/notifications.py", line 154, in make_email
    body = template.get_body()
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/jinja2/runtime.py", line 675, in __call__
    return self._invoke(arguments, autoescape)
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/flask_pluginengine/util.py", line 154, in decorator
    return func(*args, **kwargs)
  File "/home/adrian/dev/indico/src/indico/web/templates/emails/base.html", line 34, in macro

  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/flask_pluginengine/templating.py", line 108, in call
    return super(PluginJinjaContext, __self).call(__obj, *args, **kwargs)
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/jinja2/runtime.py", line 675, in __call__
    return self._invoke(arguments, autoescape)
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/flask_pluginengine/util.py", line 154, in decorator
    return func(*args, **kwargs)
  File "/home/adrian/dev/indico/src/indico/web/templates/emails/base.html", line 49, in macro

  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/flask_pluginengine/util.py", line 134, in generator
    for item in gen:
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/templates/emails/base_registration.html", line 41, in block_body

  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/flask_pluginengine/templating.py", line 108, in call
    return super(PluginJinjaContext, __self).call(__obj, *args, **kwargs)
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/jinja2/runtime.py", line 372, in __call__
    rv = concat(self._stack[self._depth](self._context))
  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/flask_pluginengine/util.py", line 134, in generator
    for item in gen:
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/templates/emails/base_registration_details.html", line 98, in block_registration_body

  File "/home/adrian/dev/indico/env-py27/lib64/python2.7/site-packages/jinja2/environment.py", line 471, in getattr
    return getattr(obj, attribute)
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/models/registrations.py", line 384, in summary_data
    _fill_from_regform()
  File "/home/adrian/dev/indico/src/indico/modules/events/registration/models/registrations.py", line 373, in _fill_from_regform
    summary[section][field] = field_summary[field]
KeyError: <RegistrationFormField(730520, registration_form_id=65936): "something">
```

**Comment by ThiefMaster** [comment]
Not compatible with the 3.2 changes (and may not even be needed anymore considering how long it stayed a draft...?)

**Comment by OmeGak** [comment]
We will still need to have a nice way to add regform fields and sections, but I understand that this is going to be possible through https://github.com/indico/indico/pull/5282.

**Comment by ThiefMaster** [comment]
Adding custom fields/sections: Not really. But the hook points I added for the cern-access plugin might do the job: https://github.com/indico/indico-plugins-cern/pull/110

**Pull request closed without merging**